stdlib/importlib/abc.pyi

```diff
@@ -82,24 +91,82 @@ if sys.version_info >= (3, 7):
         def contents(self) -> Iterator[str]: ...
 
 if sys.version_info >= (3, 9):
+    _P = TypeVar("_P", bound=Traversable)
     @runtime_checkable
     class Traversable(Protocol):
         @abstractmethod
-        def iterdir(self) -> Iterator[Traversable]: ...
+        def is_dir(self) -> bool: ...
         @abstractmethod
-        def read_bytes(self) -> bytes: ...
+        def is_file(self) -> bool: ...
         @abstractmethod
-        def read_text(self, encoding: Optional[str] = ...) -> str: ...
+        def iterdir(self: _P) -> Generator[_P, None, None]: ...
         @abstractmethod
-        def is_dir(self) -> bool: ...
+        def joinpath(self: _P, child: StrPath) -> _P: ...
```

Comment:
Maybe this should just return `Traversible`, as that would be less restrictive? If a class named `foobar.CustomPath` returns a `pathlib.PurePosixPath` when joining, it isn't a problem as far as importlib is concerned.

For future reference, we now prefer `from _typeshed import Self` for situations like this.

Reply:
Changed!